mem-ruby: Add support for CLFLUSH type instructions in MESI Three Level protocol

AmoghBhagwat opened the pull request:
Aims to fix the issue referenced [here](https://github.com/gem5/gem5/issues/2642). Correctness tests done:

- Verified that cache line is actually being flushed by measuring time of access
- Tested correctness of writing dirty data. Cache lines shared among cores are also flushed correctly. Example program:
```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <sched.h>

void clflush(void *ptr) {
    asm volatile(
        "clflush (%0)"
        :
        : "r"(ptr)
        : "memory"
    );
}

void mfence() {
    asm volatile("mfence" ::: "memory");
}

int main() {
    int data = 42;
    printf("Before clflush: data = %d\n", data);
    clflush(&data);
    printf("After clflush: data = %d\n", data);

    volatile char* x = (char*) mmap(NULL, 256*sizeof(char), PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    x[255] = 1;
    int pid = fork();
    if (pid == 0) {
        for (int i = 0; i < 128; i++) {
            x[i] = 1;
        }
        mfence();
        x[255] = 0;
        mfence();
        return 0;
    }

    while (x[255] == 1);
    clflush(&x[0]);
    clflush(&x[64]);
    int sum = 0;
    for (int i = 0; i < 128; i++) sum += x[i];

    printf("sum = %d\n", sum);

    return 0;
}

```
sum is correctly printed as 128.

Note that for this to work, these lines (279-284) need to be removed from `RubyPort.cc`. I have not included this in the PR as it would break other protocols.
```cpp
if (pkt->req->isCacheMaintenance()) {
    warn_once("Cache maintenance operations are not supported in Ruby.\n");
    pkt->makeResponse();
    schedTimingResp(pkt, curTick());
    return true;
}
```

AmoghBhagwat commented:
@erin-le any comments about the PR? The clang format test is failing because of the .sm files, perhaps a separate PR is needed for fixing it.

powerjg commented:
> erin-le any comments about the PR? The clang format test is failing because of the .sm files, perhaps a separate PR is needed for fixing it.

We shouldn't have clang-format run on .sm files. If that's what's happening here, we should fix the pre-commit and the CI test.

powerjg reviewed (commented):
lgtm. Thanks for the contribution!!!

One (hopefully small) request: If it's possible, can you run the ruby_random_test with flushes enabled with this PR? It may be that that the random tester is broken, so this may not be easily feasible. If so, just let me know.

AmoghBhagwat commented:
Updated the ruby random tester, also I have changed a few things, earlier I was writing the flushed block to memory even if it was not modified. Now it only writes back if it is dirty.

powerjg commented on src/mem/ruby/protocol/MESI_Three_Level-L0cache.sm:
```diff
@@ -152,6 +156,11 @@ machine(MachineType:L0Cache, "MESI Directory L0 Cache")
+    // Flush transitions
+    Flush, desc="Receive flush from processor";
+    Flush_Ack, desc="Flush ack from L1";
+    Flush_Data, desc="Send data to L1 during flush";
```
Nit: Could you call this something like "External_Flush_data"? From what I understand (correct me if I'm wrong). This event happens when a higher level cache receives a flush request for which this cache owns the data.

AmoghBhagwat replied:
Yes you are right, did the changes.

powerjg reviewed (commented):
Just one tiny nit. 

Thanks for doing this! Have you tested taking checkpoints with this protocol?

AmoghBhagwat commented:
> Have you tested taking checkpoints with this protocol?

I had not tested it before, seems to be working fine for `tests/gem5/checkpoint_tests/configs/x86-fs-save-checkpoint.py` and ``tests/gem5/checkpoint_tests/configs/x86-fs-restore-checkpoint.py`` by using MESI_Three_Level hierarchy, and restoring the appropriate checkpoint. However, it is not working for `x86-hello-save-checkpoint.py`, I am not sure why. I am getting the following error while creating the checkpoint. To reproduce, the lines in `RubyPort.cc` need to be removed as mentioned in the PR description.
```
src/mem/ruby/common/Consumer.cc:89: void gem5::ruby::Consumer::processCurrentEvent(): Assertion `em->clockEdge() == *curr' failed.
```

powerjg commented:
Great! Thanks. A couple of other things:

> However, it is not working for x86-hello-save-checkpoint.py, I am not sure why. I am getting the following error while creating the checkpoint. To reproduce, the lines in RubyPort.cc need to be removed as mentioned in the PR description.

I think this is OK. I have a feeling that checkpointing with Ruby doesn't work generally in SE mode. If it's works in FS mode. I'm happy!

> Note that for this to work, these lines (279-284) need to be removed from RubyPort.cc. I have not included this in the PR as it would break other protocols.

There's a new(-ish) interface in Ruby/SLICC to add metadata to protocols. See https://github.com/gem5/gem5/blob/develop/src/mem/ruby/slicc_interface/ProtocolInfo.hh and https://github.com/gem5/gem5/blob/develop/src/mem/ruby/protocol/chi/CHI.slicc#L1 for an example use case.

I think you can extend the `ProtocolInfo` object with "supports flushes" and restructure the if statement in RubyPort.cc.

AmoghBhagwat committed:
mem-ruby: Add support for CLFLUSH in MESI Three Level protocol

powerjg reviewed (dismissed):
Thanks, this is great!

erin-le commented:
@AmoghBhagwat: Hello, thanks for this PR! Could you run the following command in your `gem5` directory and push the changes? This is just to get the `clang-format-check` CI test to pass. Thank you!

```
python util/run-git-clang-format.py --ci-pr-base-commit d65257daf3f60d000cb33e4b12e79bdbeb02b878
```

AmoghBhagwat committed:
Merge branch 'develop' into mesi-three-level-clflush

AmoghBhagwat commented:
Not sure why the CI test is failing, running
`python util/run-git-clang-format.py --ci-pr-base-commit 52af36db4090ce41f5a02d207afa0cb4ae84c394`
locally does not change anything.

erin-le commented:
@AmoghBhagwat: I was able to see a change when I ran the command, but used `python3` instead of `python`, which could have made a difference. I'll push the change as a commit and see if that causes the `clang-format-check` to pass.

I also tried running the command using the commit before the first commit of the PR (d65257daf3f60d000cb33e4b12e79bdbeb02b878), but this caused `git-clang-format` to modify a file that this PR hadn't changed. I'm not sure why this happened.

erin-le committed:
mem-ruby: clang formatting for ProtocolInfo.hh
